alibi/utils/distributed.py

```diff
@@ -217,3 +224,203 @@ def _chunk(values, chunksize):
+    """
+    Inverts a permutation.
+
+    Parameters:
+    -----------
+    p
+        Some permutation of 0, 1, ..., len(p)-1. Returns an array s, where s[i] gives the index of i in p.
+
+    Returns
+    -------
+    s
+        `s[i]` gives the index of `i` in `p`.
+    """
+
+    s = np.empty_like(p)
+    s[p] = np.arange(len(p))
+    return s
+
+
+class DistributedExplainer:
+    """
+    A class that orchestrates the execution of the execution of a batch of explanations in parallel.
+    """
+    if RAY_INSTALLED:
+        import ray
+        ray = ray
+
+    def __init__(self, distributed_opts, cls, init_args, init_kwargs):
+        if not RAY_INSTALLED:
+            raise ModuleNotFoundError("Module requires ray to be installed. pip install alibi[ray] ")
+
+        self.n_jobs = distributed_opts['n_cpus']
+        self.batch_size = distributed_opts['batch_size']
+        self.algorithm = distributed_opts['algorithm']
+        self.target_fn = globals()[f"{distributed_opts['algorithm']}_target_fn"]
+        try:
+            self.post_process_fcn = globals()[f"{distributed_opts['algorithm']}_postprocess_fn"]
+        except KeyError:
+            self.post_process_fcn = None
+
+        self.explainer = cls
+        self.explainer_args = init_args
+        self.explainer_kwargs = init_kwargs
+
+        if not DistributedExplainer.ray.is_initialized():
+            DistributedExplainer.ray.init()
+
+        self.pool = self.create_parallel_pool()
+
+    def __getattr__(self, item):
+        """
+        Access to actor attributes. Should be used to retrieve only state that is shared by all actors in the pool.
+        """
+        actor = self.pool._idle_actors[0]
+        return actor.return_attribute.remote(item)
+
+    def create_parallel_pool(self):
+        """
+        Creates a pool of actors that can execute explanations in parallel.
+        """
+
+        handles = [DistributedExplainer.ray.remote(self.explainer) for _ in range(self.n_jobs)]
+        workers = [handle.remote(*self.explainer_args, **self.explainer_kwargs) for handle in handles]
+        return DistributedExplainer.ray.util.ActorPool(workers)
+
+    def batch(self, X: np.ndarray) -> enumerate:
+        """
+        Splits the input into sub-arrays according to the following logic:
+
+            - if `batch_size` is not `None`, batches of this size are created. The sizes of the batches created might \
+            vary if the 0-th dimension of `X` is not divisible by `batch_size`. For an array of length l that should be
+            split into n sections, it returns l % n sub-arrays of size l//n + 1 and the rest of size l//n.
+            - if `batch_size` is `None`, then `X` is split into `n_jobs` sub-arrays
+
+        Parameters
+        ----------
+        X
+            Array to be split.
+        Returns
+        ------
+            A list of sub-arrays of X.
+        """
+
+        n_records = X.shape[0]
+        if isinstance(X, sparse.spmatrix):
```

Comment:
This would likely never happen as the error is raised already in `KernelExplainer`?

Reply:
I guess that it's probably ok to just cast so we can split the batch and remove the `TypeError` and support the sparse matrix?

What are your thoughts on pandas? I was thinking to extract the values and just return numpy arrays with the explanations.